File: LeetCode/392. Is Subsequence/392.py
class Solution(object):
    def isSubsequence(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        
        if len(s) == 0:  # If s is empty, it is a subsequence of any string, so return true
            return True
        
        i = 0  # Initialize a pointer for s
        for c in t:  # Iterate over the characters in t
            if c == s[i]:  # If the current character in t is the same as the current character in s
                i += 1  # Increment the pointer for s
                if i == len(s):  # If the pointer for s has reached the end of s, return true
                    return True
        return False  # If the loop finishes and the pointer for s has not reached the end of s, return false
    
'''
class Solution(object):
    def isSubsequence(self, s, t):
        it = iter(t)
        return all(char in it for char in s)


        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        
'''
'''
class Solution(object):
    def isSubsequence(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        
        s_i = 0
        t_i = 0

        while s_i < len(s) and t_i < len(t):
            if t[t_i] != s[s_i]:
                t_i += 1
            else:
                t_i += 1
                s_i += 1
        
        return s_i == len(s)
'''

'''
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        i,j=0,0
        if len(s)>len(t):
            return False
        if len(s)=="":
            return True
        while i<len(s) and j<len(t):
            if s[i]==t[j]:
                i+=1
                j+=1
            else:
                j+=1
        return i==len(s)    

        
'''

File: LeetCode/392. Is Subsequence/392.java
class Solution {
    public boolean isSubsequence(String s, String t) {
        if (s.length() == 0) { // If s is empty, it is a subsequence of any string, so return true
            return true;
        }
        int indexS = 0, indexT = 0; // Initialize two pointers, one for s and one for t
        while (indexT < t.length()) { // Iterate over the characters in t
            if (t.charAt(indexT) == s.charAt(indexS)) { // If the current character in t is the same as the current character in s
                indexS++; // Increment the pointer for s
                if (indexS == s.length()) { // If the pointer for s has reached the end of s, return true
                    return true;
                }
            }
            indexT++; // Increment the pointer for t
        }
        return false; // If the loop finishes and the pointer for s has not reached the end of s, return false
    }
}
/*
 class Solution {
    public boolean isSubsequence(String s, String t) {
        int n = s.length();
        int m = t.length();

        if (n>m) return false;

        char ch;
        int j = 0;
        for (int i=0; i<n; i++) {
            ch = s.charAt(i);
            while(j<m && ch != t.charAt(j)) j++;
            if (j>=m) return false;
            j++;
        }
        return true;


    }
}
 */


// class Solution {
//     public boolean isSubsequence(String s, String t) {
//         /* Two pointers:
//             1. index of first string
//             2. index of second string

//             Linear scan over the second string
//             Increment index of first string
//             etc.
//         */

//         // Empty string
//         if (s.length() == 0) {
//             return true;
//         }

//         int i = 0;
//         for (int j = 0; j < t.length(); j++) {
//             if (t.charAt(j) == s.charAt(i)) {
//                 i++;
//             }
//             if (i >= s.length()) {
//                 return true;
//             }
//         }

//         return false;
//     }
// }